Create table SQL rewrite error with encrypt

```
[INFO ] 2021-06-03 10:56:12.015 [ShardingSphere-Command-1] ShardingSphere-SQL - Logic SQL: create table t_encrypt(user_id varchar(32), order_id varchar(32))
[INFO ] 2021-06-03 10:56:12.015 [ShardingSphere-Command-1] ShardingSphere-SQL - SQLStatement: MySQLCreateTableStatement(containsNotExistClause=false)
[INFO ] 2021-06-03 10:56:12.015 [ShardingSphere-Command-1] ShardingSphere-SQL - Actual SQL: ds_2 ::: create table t_encryptuser_cipher varchar(32),user_plain varchar(32),order_cipher varchar(32))
[ERROR] 2021-06-03 10:56:12.017 [ShardingSphere-Command-1] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'varchar(32),user_plain varchar(32),order_cipher varchar(32))' at line 1
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.Util.getInstance(Util.java:408)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:944)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3978)
```

--- a/shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/impl/EncryptCreateTableTokenGenerator.java
+++ b/shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/impl/EncryptCreateTableTokenGenerator.java
@@ -17,6 +17,7 @@
 
 package org.apache.shardingsphere.encrypt.rewrite.token.generator.impl;
 
+import com.google.common.collect.Lists;
 import org.apache.shardingsphere.encrypt.rewrite.token.generator.BaseEncryptSQLTokenGenerator;
 import org.apache.shardingsphere.encrypt.spi.EncryptAlgorithm;
 import org.apache.shardingsphere.infra.binder.statement.SQLStatementContext;
@@ -29,6 +30,7 @@ import org.apache.shardingsphere.sql.parser.sql.common.segment.ddl.column.Column
 
 import java.util.Collection;
 import java.util.LinkedList;
+import java.util.List;
 import java.util.Optional;
 
 /**
@@ -45,38 +47,44 @@ public final class EncryptCreateTableTokenGenerator extends BaseEncryptSQLTokenG
     public Collection<SQLToken> generateSQLTokens(final CreateTableStatementContext createTableStatementContext) {
         Collection<SQLToken> result = new LinkedList<>();
         String tableName = createTableStatementContext.getSqlStatement().getTable().getTableName().getIdentifier().getValue();
-        Collection<ColumnDefinitionSegment> columnDefinitionSegments = createTableStatementContext.getSqlStatement().getColumnDefinitions();
-        for (ColumnDefinitionSegment each : columnDefinitionSegments) {
+        List<ColumnDefinitionSegment> columns = Lists.newArrayList(createTableStatementContext.getSqlStatement().getColumnDefinitions());
+        for (int index = 0; index < columns.size(); index++) {
+            ColumnDefinitionSegment each = columns.get(index);
             String columnName = each.getColumnName().getIdentifier().getValue();
             Optional<EncryptAlgorithm> encryptor = getEncryptRule().findEncryptor(tableName, columnName);
             if (encryptor.isPresent()) {
-                result.addAll(getColumnTokens(tableName, columnName, each));
+                result.addAll(getColumnTokens(tableName, columnName, each, columns, index));
             }
         }
         return result;
     }
     
-    private Collection<SQLToken> getColumnTokens(final String tableName, final String columnName, final ColumnDefinitionSegment columnDefinitionSegment) {
+    private Collection<SQLToken> getColumnTokens(final String tableName, final String columnName, final ColumnDefinitionSegment column, 
+                                                 final List<ColumnDefinitionSegment> columns, final int index) {
+        boolean lastColumn = columns.size() - 1 == index;
+        int columnStopIndex = lastColumn ? column.getStopIndex() : columns.get(index + 1).getStartIndex() - 1;
         Collection<SQLToken> result = new LinkedList<>();
-        result.add(new RemoveToken(columnDefinitionSegment.getStartIndex() - 1, columnDefinitionSegment.getStopIndex() + 1));
-        result.add(getCipherColumnToken(tableName, columnName, columnDefinitionSegment));
-        getAssistedQueryColumnToken(tableName, columnName, columnDefinitionSegment).ifPresent(result::add);
-        getPlainColumnToken(tableName, columnName, columnDefinitionSegment).ifPresent(result::add);
+        result.add(new RemoveToken(column.getStartIndex(), columnStopIndex));
+        result.add(getCipherColumnToken(tableName, columnName, column, columnStopIndex));
+        getAssistedQueryColumnToken(tableName, columnName, column, columnStopIndex, lastColumn).ifPresent(result::add);
+        getPlainColumnToken(tableName, columnName, column, columnStopIndex, lastColumn).ifPresent(result::add);
         return result;
     }
     
-    private SubstitutableColumnNameToken getCipherColumnToken(final String tableName, final String columnName, final ColumnDefinitionSegment columnDefinitionSegment) {
+    private SubstitutableColumnNameToken getCipherColumnToken(final String tableName, final String columnName, final ColumnDefinitionSegment column, final int stopIndex) {
         String cipherColumn = getEncryptRule().getCipherColumn(tableName, columnName);
-        return new SubstitutableColumnNameToken(columnDefinitionSegment.getStopIndex() + 2, columnDefinitionSegment.getColumnName().getStopIndex(), cipherColumn);
+        return new SubstitutableColumnNameToken(stopIndex + 1, column.getColumnName().getStopIndex(), cipherColumn);
     }
     
-    private Optional<SubstitutableColumnNameToken> getAssistedQueryColumnToken(final String tableName, final String columnName, final ColumnDefinitionSegment columnDefinitionSegment) {
+    private Optional<SubstitutableColumnNameToken> getAssistedQueryColumnToken(final String tableName, final String columnName, final ColumnDefinitionSegment column, 
+                                                                               final int stopIndex, final boolean lastColumn) {
         Optional<String> assistedQueryColumn = getEncryptRule().findAssistedQueryColumn(tableName, columnName);
-        return assistedQueryColumn.map(optional -> new SubstitutableColumnNameToken(columnDefinitionSegment.getStopIndex() + 2, columnDefinitionSegment.getColumnName().getStopIndex(), optional));
+        return assistedQueryColumn.map(optional -> new SubstitutableColumnNameToken(stopIndex + 1, column.getColumnName().getStopIndex(), optional, lastColumn));
     }
     
-    private Optional<SubstitutableColumnNameToken> getPlainColumnToken(final String tableName, final String columnName, final ColumnDefinitionSegment columnDefinitionSegment) {
+    private Optional<SubstitutableColumnNameToken> getPlainColumnToken(final String tableName, final String columnName, final ColumnDefinitionSegment column, 
+                                                                       final int stopIndex, final boolean lastColumn) {
         Optional<String> plainColumn = getEncryptRule().findPlainColumn(tableName, columnName);
-        return plainColumn.map(optional -> new SubstitutableColumnNameToken(columnDefinitionSegment.getStopIndex() + 2, columnDefinitionSegment.getColumnName().getStopIndex(), optional));
+        return plainColumn.map(optional -> new SubstitutableColumnNameToken(stopIndex + 1, column.getColumnName().getStopIndex(), optional, lastColumn));
     }
 }

--- a/shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-rewrite/src/test/resources/encrypt/create_for_query_with_plain.xml
+++ b/shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-rewrite/src/test/resources/encrypt/create_for_query_with_plain.xml
@@ -19,6 +19,6 @@
 <rewrite-assertions yaml-rule="yaml/encrypt/query-with-plain.yaml">
     <rewrite-assertion id="create_table_for_plain">
         <input sql="CREATE TABLE t_account_bak (id int not null, name varchar(100) not null default '', password varchar(255) not null default '', primary key (`id`))" />
-        <output sql="CREATE TABLE t_account_bak (id int not null, name varchar(100) not null default '',cipher_password varchar(255) not null default '',assisted_query_password varchar(255) not null default '',plain_password varchar(255) not null default '', primary key (`id`))" />
+        <output sql="CREATE TABLE t_account_bak (id int not null, name varchar(100) not null default '', cipher_password varchar(255) not null default '', assisted_query_password varchar(255) not null default '', plain_password varchar(255) not null default '', primary key (`id`))" />
     </rewrite-assertion>
 </rewrite-assertions>

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/mix/create.xml
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/mix/create.xml
@@ -1,0 +1,40 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one or more
+  ~ contributor license agreements.  See the NOTICE file distributed with
+  ~ this work for additional information regarding copyright ownership.
+  ~ The ASF licenses this file to You under the Apache License, Version 2.0
+  ~ (the "License"); you may not use this file except in compliance with
+  ~ the License.  You may obtain a copy of the License at
+  ~
+  ~     http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<rewrite-assertions yaml-rule="yaml/mix/query-with-cipher.yaml">
+    <rewrite-assertion id="create_table_with_single_cipher_column">
+        <input sql="CREATE TABLE t_account_detail(account_id INT, password VARCHAR(100));" />
+        <output sql="CREATE TABLE t_account_detail_0(account_id INT, cipher_password VARCHAR(100), assisted_query_password VARCHAR(100), plain_password VARCHAR(100));" />
+        <output sql="CREATE TABLE t_account_detail_1(account_id INT, cipher_password VARCHAR(100), assisted_query_password VARCHAR(100), plain_password VARCHAR(100));" />
+    </rewrite-assertion>
+    <rewrite-assertion id="create_table_with_multi_cipher_column">
+        <input sql="CREATE TABLE t_account_detail(account_id INT, password VARCHAR(100), amount VARCHAR(100));" />
+        <output sql="CREATE TABLE t_account_detail_0(account_id INT, cipher_password VARCHAR(100), assisted_query_password VARCHAR(100), plain_password VARCHAR(100), cipher_amount VARCHAR(100), plain_amount VARCHAR(100));" />
+        <output sql="CREATE TABLE t_account_detail_1(account_id INT, cipher_password VARCHAR(100), assisted_query_password VARCHAR(100), plain_password VARCHAR(100), cipher_amount VARCHAR(100), plain_amount VARCHAR(100));" />
+    </rewrite-assertion>
+    <rewrite-assertion id="create_table_with_all_cipher_column">
+        <input sql="CREATE TABLE t_account_detail(password VARCHAR(100), amount VARCHAR(100));" />
+        <output sql="CREATE TABLE t_account_detail_0(cipher_password VARCHAR(100), assisted_query_password VARCHAR(100), plain_password VARCHAR(100), cipher_amount VARCHAR(100), plain_amount VARCHAR(100));" />
+        <output sql="CREATE TABLE t_account_detail_1(cipher_password VARCHAR(100), assisted_query_password VARCHAR(100), plain_password VARCHAR(100), cipher_amount VARCHAR(100), plain_amount VARCHAR(100));" />
+    </rewrite-assertion>
+    <rewrite-assertion id="create_table_with_random_blank_cipher_column">
+        <input sql="CREATE TABLE t_account_detail( account_id INT  , password VARCHAR(100) ,  amount VARCHAR(100)); " />
+        <output sql="CREATE TABLE t_account_detail_0( account_id INT  , cipher_password VARCHAR(100) ,  assisted_query_password VARCHAR(100) ,  plain_password VARCHAR(100) ,  cipher_amount VARCHAR(100), plain_amount VARCHAR(100)); " />
+        <output sql="CREATE TABLE t_account_detail_1( account_id INT  , cipher_password VARCHAR(100) ,  assisted_query_password VARCHAR(100) ,  plain_password VARCHAR(100) ,  cipher_amount VARCHAR(100), plain_amount VARCHAR(100)); " />
+    </rewrite-assertion>
+</rewrite-assertions>

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/yaml/mix/query-with-cipher.yaml
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/yaml/mix/query-with-cipher.yaml
@@ -93,6 +93,17 @@ rules:
           cipherColumn: cipher_amount
           plainColumn: plain_amount
           encryptorName: normal
+    t_account_detail:
+      columns:
+        password:
+          cipherColumn: cipher_password
+          assistedQueryColumn: assisted_query_password
+          plainColumn: plain_password
+          encryptorName: assisted_query
+        amount:
+          cipherColumn: cipher_amount
+          plainColumn: plain_amount
+          encryptorName: normal
   encryptors:
     normal:
       type: NORMAL_ENCRYPT

--- a/shardingsphere-infra/shardingsphere-infra-rewrite/shardingsphere-infra-rewrite-engine/src/main/java/org/apache/shardingsphere/infra/rewrite/sql/token/pojo/generic/SubstitutableColumnNameToken.java
+++ b/shardingsphere-infra/shardingsphere-infra-rewrite/shardingsphere-infra-rewrite-engine/src/main/java/org/apache/shardingsphere/infra/rewrite/sql/token/pojo/generic/SubstitutableColumnNameToken.java
@@ -33,14 +33,24 @@ public final class SubstitutableColumnNameToken extends SQLToken implements Subs
     
     private final String columnName;
     
+    private final boolean lastColumn;
+    
     public SubstitutableColumnNameToken(final int startIndex, final int stopIndex, final String columnName) {
         super(startIndex);
         this.stopIndex = stopIndex;
         this.columnName = columnName;
+        this.lastColumn = false;
+    }
+    
+    public SubstitutableColumnNameToken(final int startIndex, final int stopIndex, final String columnName, final boolean lastColumn) {
+        super(startIndex);
+        this.stopIndex = stopIndex;
+        this.columnName = columnName;
+        this.lastColumn = lastColumn;
     }
     
     @Override
     public String toString() {
-        return columnName;
+        return lastColumn ? ", " + columnName : columnName;
     }
 }

